src/lib.rs

```diff
@@ -150,110 +213,260 @@ impl<'object, Endian> DebugInfo<'object, Endian>
         })
     }
 
-    pub fn locate(&self, addr: u64) -> Option<(String, u64)> {
-        let mut current: Option<gimli::LineNumberRow> = None;
+    pub fn locate(&self, addr: u64) -> Option<(path::PathBuf, u64, Option<Cow<str>>)> {
+        // First, find the compilation unit for the given address
         for unit in &self.units {
             if !unit.contains_address(addr) {
                 continue;
             }
 
+            // Okay, this is the right unit. Check our DebugLine rows.
             let rows = unit.line_rows(&self.debug_line);
             if rows.is_err() {
-                continue;
+                return None;
             }
-
             let mut rows = rows.unwrap();
-            while let Ok(Some((header, row))) = rows.next_row() {
-                if row.address() > addr {
-                    if let Some(ref row) = current {
-                        let mut path = String::new();
-                        let file = row.file(header).unwrap();
-                        if let Some(directory) = file.directory(header) {
-                            let directory = directory.to_string_lossy();
-                            if !directory.starts_with("/") {
-                                if let Some(comp_dir) = unit.comp_dir() {
-                                    path.push_str(&*comp_dir.to_string_lossy());
-                                    path.push_str("/");
-                                }
-                            }
-                            path.push_str(&*directory);
-                            path.push_str("/");
-                        }
-                        path.push_str(&*file.path_name().to_string_lossy());
 
-                        return Some((path, row.line().unwrap()));
+            // Now, find the last row before a row with a higher address than the one we seek.
+            let mut current = None;
+            while let Ok(Some((_, row))) = rows.next_row() {
+                if row.address() <= addr {
+                    if row.end_sequence() {
+                        current = None;
+                    } else {
+                        // Might be the right row, but we won't know until we see the next one.
+                        // The .clone is needed so we can keep iterating
+                        current = Some(row.clone());
+                    }
+                    continue;
+                }
+                break;
+            }
+
+            // The row we just last iterated to is *after* the address, to the previous row we
+            // saw (stored in current) is the one we want. There better be one...
+            let row = current.unwrap();
```

Comment:
Despite the comment, it isn't clear to me that this `unwrap` won't panic: what if there aren't any rows in the line number program?